core/drivers/clk/clk_dt.c

```diff
@@ -0,0 +1,395 @@
+					    &len);
+		if (!compat)
+			return TEE_ERROR_GENERIC;
+
+		clk_drv = clk_get_compatible_driver(compat);
+		if (!clk_drv)
+			continue;
+
+		res = clk_drv->setup(fdt, node);
+		if (res != TEE_SUCCESS) {
+			EMSG("Failed to setup clock driver for compatible %s",
+			     compat);
+			panic();
+		} else {
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_GENERIC;
+}
+
+static void clk_probe_node(const void *fdt, int parent_node);
+
+static TEE_Result clk_probe_clock_provider_node(const void *fdt, int node)
+{
+	int len = 0;
+	int status = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	status = _fdt_get_status(fdt, node);
+	if (!(status & DT_STATUS_OK_SEC))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	/* Check if the node is a clock provider */
+	if (!fdt_getprop(fdt, node, "#clock-cells", &len))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	/* Check if node has already been probed */
+	if (clk_get_provider_by_node(node))
+		return TEE_SUCCESS;
+
+	/* Check if the node has a clock property first to setup parent */
+	res = parse_clock_property(fdt, node);
+	if (res)
+		return res;
+
+	return clk_dt_node_clock_setup_driver(fdt, node);
+}
+
+static void clk_probe_node(const void *fdt, int parent_node)
+{
+	int child = 0;
+
+	fdt_for_each_subnode(child, fdt, parent_node) {
+		clk_probe_clock_provider_node(fdt, child);
+		clk_probe_node(fdt, child);
+	}
+}
+
+static bool clk_dt_set_parent(struct clk *clk, struct clk *parent)
+{
+	size_t i = 0;
+
+	for (i = 0; i < clk_get_num_parents(clk); i++) {
+		if (clk_get_parent_by_index(clk, i) == parent) {
+			if (clk_set_parent(clk, i) != TEE_SUCCESS)
+				panic();
+			return true;
+		}
+	}
+
+	EMSG("Clock %s is not a parent of clock %s", parent->name, clk->name);
+	return false;
+}
+
+static void parse_assigned_clock(const void *fdt, int nodeoffset)
+{
+	int rate_len = 0;
+	int clock_idx = 0;
+	struct clk *clk = NULL;
+	unsigned long rate = 0;
+	struct clk *parent = NULL;
+	const uint32_t *rate_prop = NULL;
+
+	rate_prop = fdt_getprop(fdt, nodeoffset, "assigned-clock-rates",
+				&rate_len);
+	rate_len /= sizeof(uint32_t);
+
+	while (1) {
+		clk = clk_dt_get_by_idx_prop("assigned-clocks", fdt, nodeoffset,
+					     clock_idx);
+		if (!clk)
+			return;
+
+		parent = clk_dt_get_by_idx_prop("assigned-clock-parents", fdt,
+						nodeoffset, clock_idx);
+		if (parent) {
+			if (!clk_dt_set_parent(clk, parent)) {
+				EMSG("Could not set clk %s parent to clock %s",
+				     clk->name, parent->name);
+				panic();
+			}
+		}
+
+		if (rate_prop && clock_idx <= rate_len) {
+			rate = fdt32_to_cpu(rate_prop[clock_idx]);
+			if (rate && clk_set_rate(clk, rate) != TEE_SUCCESS)
+				panic();
+		}
+
+		clock_idx++;
+	}
+}
+
+static void clk_probe_assigned(const void *fdt, int parent_node)
+{
+	int len = 0;
+	int child = 0;
+	int status = 0;
+
+	fdt_for_each_subnode(child, fdt, parent_node) {
+		clk_probe_assigned(fdt, child);
+
+		status = _fdt_get_status(fdt, child);
+		if (!status)
```

Comment:
prefer explicit `if (status == DT_STATUS_DISABLED)`
or even:
```c
	fdt_for_each_subnode(child, fdt, parent_node) {
		clk_probe_assigned(fdt, child);

		if (_fdt_get_status(fdt, child) != DT_STATUS_DISABLED &&
		    fdt_getprop(fdt, child, "assigned-clocks", &len))
			parse_assigned_clock(fdt, child);
	}

Reply:
I'll use `if (status == DT_STATUS_DISABLED)`. I don't think that putting everything under the following `if` eases reading.